mockrxandroidble/src/main/java/com/polidea/rxandroidble/mockrxandroidble/RxBleClientMock.java

```diff
@@ -0,0 +1,281 @@
+         */
+        public Builder addDevice(RxBleDevice rxBleDevice) {
+            availableDevices.put(rxBleDevice.getMacAddress(), rxBleDevice);
+            return this;
+        }
+
+        /**
+         * Create the {@link RxBleClientMock} instance using the configured values.
+         */
+        public RxBleClientMock build() {
+            return new RxBleClientMock(this);
+        }
+    }
+
+    public static class DeviceBuilder {
+
+        private Integer rssi;
+        private String deviceName;
+        private String deviceMacAddress;
+        private byte[] scanRecord;
+        private RxBleDeviceServices rxBleDeviceServices;
+        private Map<UUID, Observable<byte[]>> characteristicNotificationSources;
+
+        /**
+         * Build a new {@link RxBleDevice}.
+         * <p>
+         * Calling {@link #scanRecord}, {@link #rssi} and {@link #deviceMacAddress}
+         * is required before calling {@link #build()}. All other methods
+         * are optional.
+         */
+        public DeviceBuilder() {
+            this.rxBleDeviceServices = new RxBleDeviceServices(new ArrayList<>());
+            this.characteristicNotificationSources = new HashMap<>();
+        }
+
+        /**
+         * Add a {@link BluetoothGattService} to the device.
+         *
+         * @param uuid            service UUID
+         * @param characteristics characteristics that the service should report. Use {@link CharacteristicsBuilder} to create them.
+         */
+        public DeviceBuilder addService(UUID uuid, List<BluetoothGattCharacteristic> characteristics) {
+            BluetoothGattService bluetoothGattService = new BluetoothGattService(uuid, 0);
+            for (BluetoothGattCharacteristic characteristic : characteristics) {
+                bluetoothGattService.addCharacteristic(characteristic);
+            }
+            rxBleDeviceServices.getBluetoothGattServices().add(bluetoothGattService);
+            return this;
+        }
+
+        /**
+         * Create the {@link RxBleClientMock} instance using the configured values.
+         */
+        public RxBleDeviceMock build() {
+            if (this.rssi == null) throw new IllegalStateException("Rssi required");
+            if (this.deviceMacAddress == null) throw new IllegalStateException("DeviceMacAddress required");
+            if (this.scanRecord == null) throw new IllegalStateException("ScanRecord required");
+            RxBleDeviceMock rxBleDeviceMock = new RxBleDeviceMock(deviceName,
+                    deviceMacAddress,
+                    scanRecord,
+                    rssi,
+                    rxBleDeviceServices,
+                    characteristicNotificationSources);
+
+            for (BluetoothGattService service : rxBleDeviceServices.getBluetoothGattServices()) {
+                rxBleDeviceMock.getAdvertisedUUIDs().add(service.getUuid());
```

Comment:
How do you ensure that getAdvertisedUUIDs() won't return a copy of the list? I suggest adding dedicated method to RxBleDeviceMock.